runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/control/DockerJobBundleFactory.java

```diff
@@ -0,0 +1,339 @@
+    return Platform.OTHER;
+  }
+
+  private static class SimpleStageBundleFactory<InputT> implements StageBundleFactory<InputT> {
+
+    private final BundleProcessor<InputT> processor;
+    private final ExecutableProcessBundleDescriptor processBundleDescriptor;
+
+    // Store the wrapped client in order to keep a live reference into the cache.
+    @SuppressFBWarnings private WrappedSdkHarnessClient wrappedClient;
+
+    static <InputT> SimpleStageBundleFactory<InputT> create(
+        WrappedSdkHarnessClient wrappedClient,
+        ExecutableProcessBundleDescriptor processBundleDescriptor) {
+      @SuppressWarnings("unchecked")
+      BundleProcessor<InputT> processor =
+          wrappedClient
+              .getClient()
+              .getProcessor(
+                  processBundleDescriptor.getProcessBundleDescriptor(),
+                  (RemoteInputDestination) processBundleDescriptor.getRemoteInputDestination(),
+                  wrappedClient.getStateServer().getService());
+      return new SimpleStageBundleFactory<>(processBundleDescriptor, processor, wrappedClient);
+    }
+
+    SimpleStageBundleFactory(
+        ExecutableProcessBundleDescriptor processBundleDescriptor,
+        BundleProcessor<InputT> processor,
+        WrappedSdkHarnessClient wrappedClient) {
+      this.processBundleDescriptor = processBundleDescriptor;
+      this.processor = processor;
+      this.wrappedClient = wrappedClient;
+    }
+
+    @Override
+    public RemoteBundle<InputT> getBundle(
+        OutputReceiverFactory outputReceiverFactory, StateRequestHandler stateRequestHandler)
+        throws Exception {
+      // TODO: Consider having BundleProcessor#newBundle take in an OutputReceiverFactory rather
+      // than constructing the receiver map here. Every bundle factory will need this.
+      ImmutableMap.Builder<Target, RemoteOutputReceiver<?>> outputReceivers =
+          ImmutableMap.builder();
+      for (Map.Entry<Target, Coder<WindowedValue<?>>> targetCoder :
+          processBundleDescriptor.getOutputTargetCoders().entrySet()) {
+        Target target = targetCoder.getKey();
+        Coder<WindowedValue<?>> coder = targetCoder.getValue();
+        String bundleOutputPCollection =
+            Iterables.getOnlyElement(
+                processBundleDescriptor
+                    .getProcessBundleDescriptor()
+                    .getTransformsOrThrow(target.getPrimitiveTransformReference())
+                    .getInputsMap()
+                    .values());
+        FnDataReceiver<WindowedValue<?>> outputReceiver =
+            outputReceiverFactory.create(bundleOutputPCollection);
+        outputReceivers.put(target, RemoteOutputReceiver.of(coder, outputReceiver));
+      }
+      return processor.newBundle(outputReceivers.build(), stateRequestHandler);
+    }
+
+    @Override
+    public void close() throws Exception {
+      // Clear reference to encourage cache eviction. Values are weakly referenced.
+      wrappedClient = null;
```

Comment:
Wouldn't the StageBundleFactory itself become unreferenced anyway shortly after someone called close() on it?

Reply:
Possibly, but not for sure. It's possible that clients call close but maintain references to this `SimpleStageBundleFactory`. I think we're less likely to leak unused environments this way, but I could be convinced otherwise.